### ISSUE TITLE
gulp-istanbul loads files with persistent connections

### ISSUE THREAD

**Issue by moxious** [SBoudrias/gulp-istanbul#88]
In gulpfile.js, I have a setup that looks like this:

```
    return gulp.src(config.sourceCode)
        .pipe(istanbul({ includeUntested: true }))
        .pipe(istanbul.hookRequire());
```

An issue I'm running into is that some of my modules may need to create and maintain long-running connections (e.g. to a database); these modules provide shutdown methods, but istanbul seems to be requiring or using the modules in a way that is in parallel to the rest of my testing process; shutdown is being called within the test, but istanbul seems to have a separate copy open, is this possible?   At the end of the test routine, gulp simply hangs, owing to those modules keeping open connections.

Now, I can avoid this by omitting modules that behave this way from the `gulp.src` but then I have to give up test coverage metrics on those files, which I'd rather not.  

Any suggestions on what istanbul is doing here that would be keeping these modules open?

**Comment by SBoudrias** [comment]
Istanbul is hooking the require calls to return the instrumented versions. So yeah, inquirer will returns its own version of the file.

Now why your program doesn't always get the covered version, that I don't know. But you're probably doing something funky with requires.

**Comment by moxious** [comment]
Is there a way with a hook to gain access to whichever version it is requiring so that I can call the appropriate shutdown methods when tests are complete?

**Comment by SBoudrias** [comment]
@moxious the hooks acts just like normal require. If you require your module and call shutdown, it'll just work.

**Comment by moxious** [comment]
That's what I would have thought too...we have a gulp on `'finish'` event we're listening for where we require the modules that need shutdown and then do that; my experience though is that if I include one of those modules in istanbul's coverage domain, they're getting "re-required" (I'm not sure that even entirely makes sense)....e.g. I see errors like this:

```
events.js:154
      throw er; // Unhandled 'error' event
      ^
Error: A model with this name has already been registered.
    at Ottoman._buildAndRegisterModel (/Users/mda/sw/ow-back/node_modules/ottoman/lib/ottoman.js:706:11)
    at Ottoman.model (/Users/mda/sw/ow-back/node_modules/ottoman/lib/ottoman.js:779:15)
    at Object.<anonymous> (/Users/mda/sw/ow-back/models/portfolio.js:9:423)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/mda/sw/ow-back/node_modules/istanbul/lib/hook.js:107:24)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (/Users/mda/sw/ow-back/models/index.js:9:1140)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions.(anonymous function) [as .js] (/Users/mda/sw/ow-back/node_modules/istanbul/lib/hook.js:107:24)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
```

This is the ODM (Ottoman) complaining that the models are being re-defined, which is what I would expect to happen if the file had been required in two separate processes, except we're not spawning those.   Now, I can get rid of this error and also make shutdown work correctly if I exclude the module in question from istanbul, otherwise I either get errors like the above, or, (in the case of another module) it requires the module twice, opens two sets of connections to redis; when the tests end, one set of connections get closed gracefully by my shutdown handler, the other set just dangles and the process doesn't terminate.

**Comment by moxious** [comment]
Digging through the gulp file, I found the issue - a dependent task of test was doing some default data fixture insert stuff; in essence by implication, an earlier dependent task was requiring the module that istanbul was later hookRequire'ing.   This was confusing the system I believe; I was able to resolve some of these issues by getting the other gulp tasks (which required the long-running modules in question) out of the way and by pushing up the istanbul hook process as early as I could.

I don't think that's the only issue (because I'm still experiencing hangs) but it's at least part of it.
